pkg/reconcilers/operator/discoveryservice/generators/deployment.go

```diff
@@ -160,17 +160,7 @@ func (cfg *GeneratorOptions) Deployment(hash string) func() *appsv1.Deployment {
 					},
 				},
 				Strategy: appsv1.DeploymentStrategy{
-					Type: appsv1.RollingUpdateDeploymentStrategyType,
-					RollingUpdate: &appsv1.RollingUpdateDeployment{
-						MaxUnavailable: &intstr.IntOrString{
-							Type:   intstr.String,
-							StrVal: "25%",
-						},
-						MaxSurge: &intstr.IntOrString{
-							Type:   intstr.String,
-							StrVal: "25%",
-						},
-					},
+					Type: appsv1.RecreateDeploymentStrategyType,
```

Comment:
Isn't it dangerous?

What if the new deployment is wrong? it will first kill a working pod, then create a new one that might be wrong

Is it not better to use just the lock mechanism, so:
- old pod is the leader
- new pod gets created, still not passing healthchecks, still not leader
- new pods is passing healthcheck, leader election starts
- old pod gets terminated, while new pod becomes leader


That way you can ensure that there is always only 1 running pod which is correct passing healthchecks

WDYT?

Reply:
The lock mechanism will only work on future releases, not for this upgrade because the current deployment doesn't have lock enabled. Deployment strategy fixes the current upgrade process, adding the lock is the mechanism that will solve it for future releases. We could go back to a rollout strategy in the future.

Reply:
ok fair enough!

let's maybe open an issue to remember in the future to remove this replacement strategy?

Reply:
so, maybe the problem is the current readiness healthcheck?

How operator-sdk manage to execute the reconcile workers only on the new created pod once the leader is elected?

Reply:
It's all controller-runtime logic. It only runs the controller/s in the pod that hold the lock, but the health check is completely unrelated to the lock mechanism and returns ok anyway even if the pod doesn't hold the lock.
A custom health check could be added that only passes if the pod holds the lock, but then at some point the liveness check will get the pod killed because it's not passing the health check... so there's a new whole level of problems with that approach. Maybe using the current health check as liveness and the custom one that only returns ok if the pod holds the lock as the readiness could do the trick. I'll open an issue to explore that option.

Replacement strategy isn't that bad anyway, it's very very unlikely that the new Deployment doesn't pass the health checks. I can only think of it being a problem if the image pull fails... but it's highly unlikely because it's the same image that the operator uses, which at the point the discovery service rollout is already running the new version.